pkg/reconciler/git/client.go

```diff
@@ -26,6 +26,81 @@ func (d *Client) Worktree() (*git.Worktree, error) {
 	return d.repo.Worktree()
 }
 
-func (d *Client) ResolveRevision(rev gitp.Revision) (*gitp.Hash, error) {
+func (d *Client) ResolveRevisionOrBranchHead(rev gitp.Revision) (*gitp.Hash, error) {
+	// no version provided. We will set rev to the default branch name.
+	if rev.String() == "" {
+		branch, err := d.DefaultBranch()
+		if err != nil {
+			return nil, err
+		}
+		rev = gitp.Revision(branch.Name().Short())
+	}
+	// is rev a branch name? We can't use repo.Branch() because that checks local branches only :/
+	// So, we try to resolve rev as a remote branch reference. If that works, we
+	// can just use it 's head!
+	branchRev, err := d.repo.ResolveRevision(
+		gitp.Revision(
+			fmt.Sprintf("refs/remotes/origin/%s", rev.String()),
+		),
+	)
+	if err != nil && err != gitp.ErrReferenceNotFound {
+		return nil, err
+	}
+	if branchRev != nil { // this is a branch
+		return d.repo.ResolveRevision(gitp.Revision(branchRev.String()))
+	}
 	return d.repo.ResolveRevision(rev)
 }
+
+func (d *Client) Fetch(path string, o *git.FetchOptions) error {
+	var err error
+	if d.repo == nil {
+		d.repo, err = git.PlainOpen(path)
+		if err != nil {
+			return err
+		}
+	}
+	err = d.repo.Fetch(o)
+	if err == git.NoErrAlreadyUpToDate {
+		return nil
+	}
+	return err
+}
+
+func (d *Client) PlainCheckout(path string, o *git.CheckoutOptions) error {
+	var err error
+	if d.repo == nil {
```

Comment:
What the user of this method should expect if he will have both: `a repository` and a `a path`?

Reply:
Repository will take precedence. Since the the repository is not passed, it's only initialized if this client instance is used to clone the repo from upstream. In which case we don't need to "open" it again.

Reply:
would recommend to use the Go best-practise by using the `New...` fucntion which allows to specify the repo-path just once. This will simplify the usage of the API as it's clear that the path can be defined only once and the repo will be coupled to this path.